File: Models/Room.cs

using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Mvc.ModelBinding;

namespace RoomMonitor.Models
{

    public abstract class BaseRoomKey : BaseHotelKey
    {

        // public override string HotelChain { get; set; }
        // public override string CountryCode { get; set; }
        // public override string Town { get; set; }
        // public override string Suburb { get; set; }

        [StringLength(100)]
        public abstract string RoomType { get; set; }

        [StringLength(100)]
        public abstract string Name { get; set; }
    }

    public class RoomKey : BaseRoomKey
    {
        [BindRequired]
        public override string HotelChain { get; set; }

        [BindRequired]
        public override string CountryCode { get; set; }

        [BindRequired]
        public override string Town { get; set; }

        [BindRequired]
        public override string Suburb { get; set; }

        [BindRequired]
        public override string RoomType { get; set; }

        [BindRequired]
        public override string Name { get; set; }

    }

    public class Room : BaseRoomKey
    {
        public override string HotelChain { get; set; }
        public override string CountryCode { get; set; }
        public override string Town { get; set; }
        public override string Suburb { get; set; }

        [Required]
        public override string RoomType { get; set; }

        [Required]
        public override string Name { get; set; }

        [Required]
        public List<string> Tags { get; set; }

        public RoomKey GetKey() => new RoomKey
        {
            HotelChain = HotelChain,
            CountryCode = CountryCode,
            Town = Town,
            Suburb = Suburb,
            RoomType = RoomType,
            Name = Name
        };

        public void SetHotelKey(HotelKey hotelKey)
        {
            HotelChain = hotelKey.HotelChain;
            CountryCode = hotelKey.CountryCode;
            Town = hotelKey.Town;
            Suburb = hotelKey.Suburb;
        }

        public void SetHotelKey(RoomKey roomKey)
        {
            HotelChain = roomKey.HotelChain;
            CountryCode = roomKey.CountryCode;
            Town = roomKey.Town;
            Suburb = roomKey.Suburb;
        }
    }
}

File: ClientApp/src/app/shared/models/Room.ts
import { IDeserializable } from "../interfaces/IDeserializable";
import { ICompositeKey } from "../interfaces/ICompositeKey";
import { IHotelKey } from "src/app/shared/models/Hotel";

export interface IRoomKey extends IHotelKey {
  roomType: string;
  name: string;
}

export interface IRoom extends IRoomKey {
  tags: string[];
}

export class Room
  implements IRoom, IDeserializable<IRoom>, ICompositeKey<IRoomKey> {
  hotelChain: string;
  countryCode: string;
  town: string;
  suburb: string;
  roomType: string;
  name: string;
  tags: string[];

  deserialize(input: IRoom): this {
    return Object.assign(this, input);
  }

  getKey(): IRoomKey {
    return {
      hotelChain: this.hotelChain,
      countryCode: this.countryCode,
      town: this.town,
      suburb: this.suburb,
      roomType: this.roomType,
      name: this.name,
    };
  }
}
